pkg/splunk/enterprise/secrets.go

```diff
@@ -0,0 +1,293 @@
+package enterprise
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	splcommon "github.com/splunk/splunk-operator/pkg/splunk/common"
+	splctrl "github.com/splunk/splunk-operator/pkg/splunk/controller"
+)
+
+// GetNamespaceScopedSecret retreives namespace scoped secret
+func GetNamespaceScopedSecret(c splcommon.ControllerClient, namespace string) (*corev1.Secret, error) {
+	var namespaceScopedSecret corev1.Secret
+
+	// Check if a namespace scoped secret exists
+	namespacedName := types.NamespacedName{Namespace: namespace, Name: GetNamespaceScopedSecretName(namespace)}
+	err := c.Get(context.TODO(), namespacedName, &namespaceScopedSecret)
+	if err != nil {
+		// Didn't find it
+		return nil, err
+	}
+
+	return &namespaceScopedSecret, nil
+}
+
+// GetVersionedSecretVersion checks if the secretName includes the versionedSecretIdentifier and if so, extracts the version
+func GetVersionedSecretVersion(secretName string, versionedSecretIdentifier string) (int, error) {
+	// Extracting version from secret's name
+	version := strings.TrimPrefix(secretName, GetVersionedSecretName(versionedSecretIdentifier, ""))
+
+	// Check if the secretName includes the versionedSecretIdentifier
+	if version != secretName {
+		// Includes, version extracted, check if version number is valid
+		versionInt, err := strconv.Atoi(version)
+		if err != nil {
+			return -1, errors.New(nonIntegerVersionError)
+		}
+
+		// Versions should be > 0
+		if versionInt <= 0 {
+			return -1, errors.New(lessThanOrEqualToZeroVersionError)
+		}
+
+		return versionInt, nil
+	}
+
+	// Secret name not matching required criteria
+	return -1, fmt.Errorf(nonMatchingStringError, secretName, versionedSecretIdentifier)
+}
+
+// GetExistingLatestVersionedSecret retreives latest EXISTING versionedSecretIdentifier based secret existing currently in the namespace
+func GetExistingLatestVersionedSecret(c splcommon.ControllerClient, namespace string, versionedSecretIdentifier string) (*corev1.Secret, int) {
+	scopedLog := log.WithName("GetExistingLatestVersionedSecret").WithValues(
+		"versionedSecretIdentifier", versionedSecretIdentifier,
+		"namespace", namespace)
+
+	// Get list of secrets in K8S cluster
+	secretList := corev1.SecretList{}
+
+	// Retreive only secrets only from namespace
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
+	}
+
+	err := c.List(context.TODO(), &secretList, listOpts...)
+	if err != nil || len(secretList.Items) == 0 {
+		scopedLog.Error(err, "Secrets not found in namespace")
+		return nil, -1
```

Comment:
Nit: can be existingLatestVersion

Reply:
I thought of avoiding creation of a variable existingLatestVersion if there is no secret in the namespace to save memory?